File: solutions/php/proverb/1/Proverb.php
<?php
declare(strict_types=1);

class Proverb
{
    public function recite(array $words): array
    {
        $result = [];
        foreach ($words as $i => $word) {
            if ($i === count($words) - 1) {
                $result[] = "And all for the want of a {$words[0]}.";
                return $result;
            }
            $result[] = "For want of a $word the {$words[$i + 1]} was lost.";
        }
        return $result;
    }
}

File: solutions/go/proverb/1/proverb.go
package proverb

import "fmt"

const wantFormat = "For want of a %s the %s was lost."
const allFormat = "And all for the want of a %s."

func Proverb(rhyme []string) []string {
	if len(rhyme) == 0 {
		return []string{}
	}

	result := make([]string, 0, len(rhyme))
	for i := 0; i < len(rhyme)-1; i++ {
		result = append(result, fmt.Sprintf(wantFormat, rhyme[i], rhyme[i+1]))
	}
	result = append(result, fmt.Sprintf(allFormat, rhyme[0]))
	return result
}
